{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# script adapted from jeremy roop\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import Bio\n",
    "from Bio import SeqIO\n",
    "import os\n",
    "import csv\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pi(seq_dict):\n",
    "\n",
    "    total_sites = 0\n",
    "    diff_sites = 0.0\n",
    "    strains_run = []\n",
    "\n",
    "    for i in range(len(seq_dict)):\n",
    "        strain1 = seq_dict[i][0]\n",
    "        strain1_seq = seq_dict[i][1]\n",
    "        seq_length = len(strain1_seq)\n",
    "        \n",
    "        for j in range(i+1, len(seq_dict)):\n",
    "            \n",
    "            strain2 = seq_dict[j][0]\n",
    "            strain2_seq = seq_dict[j][1]\n",
    "\n",
    "            for pos, nuc in enumerate(strain1_seq):\n",
    "\n",
    "                alleles = set([nuc.upper(), strain2_seq[pos].upper()])\n",
    "                \n",
    "                print(alleles)\n",
    "\n",
    "                if \"N\" in alleles or '-' in alleles:\n",
    "                    continue\n",
    "\n",
    "                if len(alleles) > 1:\n",
    "                    \n",
    "                    print('here')\n",
    "                    \n",
    "                    diff_sites+=1\n",
    "\n",
    "                total_sites+=1\n",
    "                \n",
    "            if total_sites == 0:\n",
    "                                \n",
    "                break\n",
    "\n",
    "    if total_sites == 0:\n",
    "\n",
    "        return None     \n",
    "\n",
    "#             strains_run += [strain1+\"-\"+strain2]\n",
    "\n",
    "#         if len(strains_run) == 0:  ## if for a given gene no comparison was made due to there being no sequence data\n",
    "#             continue\n",
    "\n",
    "    print(diff_sites, total_sites)\n",
    "    return (diff_sites/total_sites)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#run on all spar strains\n",
    "\n",
    "results = {}\n",
    "\n",
    "directory = \"/Users/clairedubin/spur/peroxisomes/alignments/scer_spar_by_scer_population_091919/1WineEuropean/\"\n",
    "\n",
    "all_files = os.listdir(directory)  \n",
    "\n",
    "for i, file in enumerate(all_files):\n",
    "    \n",
    "\n",
    "    start_time = time.time()\n",
    "\n",
    "    print(i, file)\n",
    "\n",
    "    gene = file.split('_')[3]\n",
    "\n",
    "    seq_dict = []\n",
    "\n",
    "    for record in SeqIO.parse(directory+file, \"fasta\"):\n",
    "\n",
    "        if record.id == 'Spar;':\n",
    "\n",
    "            if 'N-44' not in record.description and 'YPS138' not in record.description:\n",
    "\n",
    "                seq_dict += [(record.description, record.seq)]   \n",
    "\n",
    "    if len(seq_dict) > 8:\n",
    "\n",
    "        result = pi(seq_dict)\n",
    "        print(gene, result)\n",
    "        results[gene] = result\n",
    "\n",
    "        with open('/Users/clairedubin/spur/peroxisomes/pi/pi_eurospar_032620.csv', 'a') as f:\n",
    "            \n",
    "            writer = csv.writer(f)\n",
    "            writer.writerow([gene, result])\n",
    "\n",
    "            f.close()\n",
    "\n",
    "    end_time = time.time()\n",
    "\n",
    "    print(\"total time taken this loop: \", end_time - start_time)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
